Allow to record historical prices (#3150)
rotki/rotki: docs/api.rst

   :statuscode 500: Internal rotki error
   :statuscode 502: An external service used in the query such as cryptocompare/coingecko could not be reached or returned unexpected response.

.. http:put:: /api/(version)/assets/prices/historical

    Manually adds the price of an asset against another asset at a certain timestamp to the database.

   .. note::

>       This endpoint can also be queried asynchronously by using ``"async_query": true``.

Reviewer: Why would you make this call aynchronous? Isn't it just a DB query? For such calls we don't offer asynchronous option as they should be blazing fast and not need an async call.

Author: I thought async queries should be used with db always

Reviewer: Nope. I don't understand what gave you this idea. Async queries are used for anything that may not be fast/almost instant. Which means ... pretty much anything that has to do with remote queries or really long calculations.

A local DB query is one of the few things that are fast. For example check the tag db queries.

      PUT /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
            "from_asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
            "to_asset": "USD",
            "async_query": true,
            "timestamp": 1611166335,
            "price": "1.20"
       }

   :reqjson string from_asset: Asset for what the price is given.
   :reqjson string to_asset: The asset against which the price is given.
   :reqjson int timestamp: The unix timestamp for which to save the price
   :reqjson string price: Price at the timestamp given.

   **Example Response**:

   .. sourcecode:: http

      HTTP/1.1 200 OK
      Content-Type: application/json

      {
          "result": true,
          "message": ""
      }

   :resjson object result: Always true
   :statuscode 200: Operation sent to database.
   :statuscode 400: Provided JSON is in some way malformed.
   :statuscode 500: Internal rotki error

.. http:get:: /api/(version)/assets/prices/historical

    Queries prices of an asset against another asset at a certain timestamp to the database.

   .. note::

>       This endpoint can also be queried asynchronously by using ``"async_query": true``.

Reviewer: Same here. Why async?

   **Example Request**:

   .. http:example:: curl wget httpie python-requests

      GET /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
          "asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620"
       }

>    :reqjson string asset: Asset for what the price is retrieved.

Reviewer: Specify from_asset here. Also allow filtering by `to_asset`.

```suggestion
   :reqjson string from_asset: The from_asset for which the price is retrieved.
```

   **Example Response**:

   .. sourcecode:: http

      HTTP/1.1 200 OK
      Content-Type: application/json

      {
          "result": [
            {
              "from_asset": "_ceth_0xD533a949740bb3306d119CC777fa900bA034cd52",
              "to_asset": "USD",
              "timestamp": 1611166335,
              "price": "1.20"
            }, 
            {
              "from_asset": '_ceth_0xD533a949740bb3306d119CC777fa900bA034cd52',
              "to_asset": 'USD',
              "timestamp": 1611166340,
              "price": "1.40"
            }
          ],
          "message": ""
      }

   :resjson object result: List with information for each historical price.
   :statuscode 200: Operation executed.
   :statuscode 400: Provided information is in some way malformed.
   :statuscode 500: Internal rotki error

.. http:delete:: /api/(version)/assets/prices/historical

    Deletes price of an asset against another asset at a certain timestamp from the database.

   .. note::

>       This endpoint can also be queried asynchronously by using ``"async_query": true``.

Reviewer: Async query is not required.

   :statuscode 500: Internal rotki error
   :statuscode 502: An external service used in the query such as cryptocompare/coingecko could not be reached or returned unexpected response.

.. http:put:: /api/(version)/assets/prices/historical

    Manually adds the price of an asset against another asset at a certain timestamp to the database.

   **Example Request**:

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
            "from_asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
            "to_asset": "USD",

>             "async_query": true,

Reviewer: Argument should no longer exist.

```suggestion
```

            "price": "1.20"
       }

   :reqjson string from_asset: The asset for which the price is given.
   :reqjson string to_asset: The asset against which the price is given.
   :reqjson int timestamp: The unix timestamp for which to save the price
   :reqjson string price: Price at the timestamp given.

   **Example Response**:

   .. sourcecode:: http

      HTTP/1.1 200 OK
      Content-Type: application/json

      {
          "result": true,
          "message": ""
      }

   :resjson object result: Always true
   :statuscode 200: Operation sent to database.
   :statuscode 400: Provided JSON is in some way malformed.
   :statuscode 500: Internal rotki error

.. http:patch:: /api/(version)/assets/prices/historical

    Edits price for a manually added price if it already exists in the database. Returns false
    if no entry was updated.

   **Example Request**:

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
            "from_asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
            "to_asset": "USD",

>             "async_query": true,

Reviewer: Argument should no longer exist.

```suggestion
```

              "price": "1.40"
            }
          ],
          "message": ""
      }

   :resjson object result: List with information for each historical price.
   :statuscode 200: Operation executed.
   :statuscode 400: Provided information is in some way malformed.
   :statuscode 500: Internal rotki error

.. http:delete:: /api/(version)/assets/prices/historical

    Deletes price of an asset against another asset at a certain timestamp from the database.

   **Example Request**:

   .. http:example:: curl wget httpie python-requests

      DELETE /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
        "from_asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
        "to_asset": "USD",
        "timestamp": 1611166335
       }

   :reqjson string from_asset: The asset for which the price is given.
   :reqjson string to_asset: The asset against which the price is given.
   :reqjson int timestamp: The unix timestamp for which to save the price

   **Example Response**:

   .. sourcecode:: http

      HTTP/1.1 200 OK
      Content-Type: application/json

      {

>         "result": null,

Reviewer: Should this be true for success and false for error?

   :statuscode 500: Internal rotki error
   :statuscode 502: An external service used in the query such as cryptocompare/coingecko could not be reached or returned unexpected response.

.. http:put:: /api/(version)/assets/prices/historical

    Saves in the database the price for an asset in a given currency at a certain timestamp.

   .. note::
      This endpoint can also be queried asynchronously by using ``"async_query": true``.

   **Example Request**:

   .. http:example:: curl wget httpie python-requests

      PUT /api/1/assets/prices/historical HTTP/1.1
      Host: localhost:5042
      Content-Type: application/json;charset=UTF-8

       {
            "from_asset": "_ceth_0xD71eCFF9342A5Ced620049e616c5035F1dB98620",
            "to_asset": "USD",
            "async_query": true,
            "timestamp": 1611166335,
            "price": "1.20"
       }

>    :reqjson string from_asset: Asset for what the price is given.

Reviewer: This is not fixed.
